### PR TITLE
Add ability to register as OpenHIM mediator and use mediator config to configure mappings

### PR DESCRIPTION
**Pull request by rcrichton**

### CODE REVIEW

File: create-test-tables.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
+curl -v 'http://localhost:8124/' --data-binary @- << EOF
+  CREATE TABLE patient(
+    id String NULL,
```
Comment (marrouchi) [inline create-test-tables.sh:3]: Might be better to choose either snake or camel case for the attribute names. There seems to be a tendency towards writing identifiers in snake case in SQL.

File: src/index.ts [hunk 2]
```diff
@@ -1,89 +1,148 @@
...
+  alterRowIntoClickhouse
+} from './clickhouse/utils'
+import mediatorConfig from './data/mediator-config.json'
+
+let fhirMappings: FhirMapping[] = require('./data/fhir-mapping.json')
+// Set default mediator FHIR mapping from config file
+mediatorConfig.config.fhirMappings = JSON.stringify(fhirMappings, null, 2)
 
-const plugins: Map<string, FhirPlugin> = GetFhirPlugins(fhirMappings);
+const plugins: Map<string, FhirPlugin> = GetFhirPlugins(fhirMappings)
```
Comment (marrouchi) [inline src/index.ts:25]:
```suggestion
const plugins: Map<string, FhirPlugin> = getFhirPlugins(fhirMappings)
```

File: src/index.ts [hunk 3]
```diff
@@ -1,89 +1,148 @@
...
+    console.error('Invalid fhir-mapping.json file')
+    fhirMappingValidationErrors.forEach(error => console.error(error))
+    process.exit(1)
+  }
+
+  const topics = fhirMappings.map(fhirMapping =>
+    fhirMapping.resourceType.toLowerCase()
+  )
+
+  await consumer.connect()
```
Comment (marrouchi) [inline src/index.ts:88]: We may need to try and catch here ?

File: src/index.ts [hunk 4]
```diff
@@ -1,89 +1,148 @@
...
 
-  await consumer.connect();
-  await producer.connect();
-  await consumer.subscribe({ topics, fromBeginning: true });
+    const config = await promisify(fetchConfig)(openhimAuth)
+    // set config to that received from the OpenHIM
+    fhirMappings = JSON.parse(config.fhirMappings)
+  }
+
+  const fhirMappingValidationErrors = ValidateFhirMappingsJson(fhirMappings)
```
Comment (marrouchi) [inline src/index.ts:77]:
```suggestion
  const fhirMappingValidationErrors = validateFhirMappingsJson(fhirMappings)
```

File: src/index.ts [hunk 5]
```diff
@@ -1,89 +1,148 @@
...
+import {
+  loadDataIntoClickhouse,
+  selectByIdClickhouse,
+  alterRowIntoClickhouse
+} from './clickhouse/utils'
+import mediatorConfig from './data/mediator-config.json'
+
+let fhirMappings: FhirMapping[] = require('./data/fhir-mapping.json')
+// Set default mediator FHIR mapping from config file
+mediatorConfig.config.fhirMappings = JSON.stringify(fhirMappings, null, 2)
```
Comment (marrouchi) [inline src/index.ts:23]: I may be wrong, I wonder why we need to stringify the fhirMapping. Also, I saw that in openHIM there's the concept of variables, I wonder we re-use of that ?

File: src/index.ts (hunk 4 above)
Comment (marrouchi) [reply src/index.ts:77]: At this line we are not sure if fhirMappings is not empty (shouldRegisterMediator = false && newFhirMappings = false)

File: src/index.ts [hunk 6]
```diff
@@ -1,89 +1,148 @@
...
-      console.log(`- ${prefix} ${message.key}#${message.value}`);
+    eachMessage: async ({topic, partition, message}) => {
+      const prefix = `${topic}[${partition} | ${message.offset}] / ${message.timestamp}`
+      console.log(`- ${prefix} ${message.key}#${message.value}`)
 
-      const entry: Entry = JSON.parse(message.value?.toString() ?? "");
+      const entry: Entry = JSON.parse(message.value?.toString() ?? '')
 
-      const tableMappings: Table[] = GetTableMappings(fhirMappings, entry, plugins);
+      const tableMappings: Table[] = GetTableMappings(
```
Comment (marrouchi) [inline src/index.ts:98]:
```suggestion
      const tableMappings: Table[] = getTableMappings(
```

Comment (marrouchi) [inline src/index.ts:98]:
```suggestion
      const tableMappings: TableMapping[] = GetTableMappings(
```

File: src/index.ts [hunk 7]
```diff
@@ -1,89 +1,148 @@
...
+        fhirMappings,
+        entry,
+        plugins
+      )
 
       for (const tableMapping of tableMappings) {
-        const res = await selectByIdClickhouse(tableMapping);
+        const res = await selectByIdClickhouse(tableMapping)
```
Comment (marrouchi) [inline src/index.ts:105]: Do we need a try/catch here ?

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: create-test-tables.sh (hunk 1 above)
Comment (rcrichton) [reply create-test-tables.sh:3]: Unfortunately this has to conform to the identifiers used in the default `fhir-mapping.json` file. I'm not sure it's the scope of this PR to change that default mapping as well.

File: src/index.ts (hunk 5 above)
Comment (rcrichton) [reply src/index.ts:23]: The data type that I'm using for storing the config in the OpenHIM is a simple string. So, this has to be converted to a string here else the validation will complain in the OpenHIM config API.

File: src/index.ts (hunk 2 above)
Comment (rcrichton) [reply src/index.ts:25]: This was an existing function. I also don't like function starting with captal unless they are in classes but this occurs a few times in this repo and I think should be corrected separately.

Comment (rcrichton) [reply src/index.ts:25]: I decided to resolve this anyways.

File: src/index.ts (hunk 4 above)
Comment (rcrichton) [reply src/index.ts:77]: It won't be empty as it is always set to the default of what is in the default config file. See line 21.

File: src/index.ts (hunk 3 above)
Comment (rcrichton) [reply src/index.ts:88]: Anything that throws in the function will get caught on line 121 and the error will be printed to the console. Are you thinkoing of any other error handling. For me that seemed enough, as there is nothing else custom that I'd want to do.

Comment (rcrichton) [reply src/index.ts:88]: Anything that throws in the function will get caught on line 121 and the error will be printed to the console. Are you thinkoing of any other error handling. For me that seemed enough, as there is nothing else custom that I'd want to do.

File: src/index.ts (hunk 7 above)
Comment (rcrichton) [reply src/index.ts:105]: Anything that throws in the function will get caught on line 121 and the error will be printed to the console. Are you thinkoing of any other error handling. For me that seemed enough, as there is nothing else custom that I'd want to do.

File: README.md [hunk 8]
```diff
@@ -11,6 +12,24 @@ Run `yarn`
 
 Run `yarn start`
 
+### Launch dev dependencies
+
+To test locally you will need a few dependencies: Kafka, Clickhouse and OpenHIM. Use the following command to setup those in a dev environment.
```
Comment (michaelloosen) [inline README.md:17]: Curious to know why you set it up outside of the platform and not suggesting it's tested with platform dev mode?

**Review by marrouchi — APPROVED** [review]

File: README.md (hunk 8 above)
Comment (rcrichton) [reply README.md:17]: I figured this was the easiest and quickest way to launch the dependencies. It starts in a few seconds as it skips all the checks and config that the platform does. I think the docker compose way still makes sense for testing single components but for testing full solutions the platform is a more complete tool.

**Review by michaelloosen — DISMISSED** [review]

**Pull request merged**